### PR TITLE
Use abstract object for retreat

### PR DESCRIPTION
**Pull request by RoiEXLab**
Follow-up to #5808
Couldn't use optional, because subs already use the current territory with a different meaning, and also couldn't move this object a layer outside to prevent special null meaning magic because I didn't want to break network compatibility for games.

## Functional Changes

[] New map or map update
[] New Feature
[] Feature update or enhancement
[] Feature Removal
[x] Code Cleanup or refactor
[] Configuration Change
[] Problem fix:  
[] Other:   

## Testing

[] Manual testing done

### CODE REVIEW

**Comment by DanVanAtta** [comment]
@asvitkine I think you might be the best reviewer for this update. Please let me know if you won't have an opportunity to take a look at this update in the next couple days.

File: game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java [hunk 1]
```diff
@@ -375,8 +367,33 @@ private Territory getSubmerge(final String message) {
     return retreatTo.get();
   }
 
-  private boolean showSubmergeDialog(
-      final String message, final AtomicReference<Territory> retreatTo) {
+  private Action getPlayerAction(
```
Comment (asvitkine) [inline game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: This still feels a bit clunky - since it's not just some random player action, but something that returns a RetreatResult.

How about just not having this method and inlining it into getRetreat() instead?

e.g.

```
final String title;
final Suppler<RetreatResult> showDialog;
if (!submerge || possible.size() > 1) {
  title = "Retreat?";
  showDialog = () -> showSubmergeDialog(message), retreatTo, latch);
} else {
  title = "Submerge Subs?";
  showDialog = () -> showSubmergeDialog(message);
}
```

And then the inlined SwingAction.of() call.

Otherwise it's hard to follow what's going on especially since there is a lot of stuff happening in this class already - so adding all these new methods just introduces more complexity.

Another option is to move the logic into separate classes for the logic, but not sure of the best set up there...

... If we step back a bit, even with this refactoring, I feel that this classes suffers a lot from mixing game logic with Swing UI code. Ultimately we'd probably want to recode things such that low level concepts - like use of JOptionPane - are abstracted away, but the business logic remains. But that's probably too big of a change for what you were after here.

Another though - I wonder if the Java Future interface would be better suited here than the combination of CountDownLatch and AtomicReference. Doesn't Future provide exactly what we need?

Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: Using future is an excellent idea, I'll try that.

**Comment by RoiEXLab** [comment]
@asvitkine Updated

File: game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java [hunk 2]
```diff
@@ -340,43 +349,66 @@ void notifyRetreat(final Collection<Unit> retreating) {
...
+    final String title;
+    final Supplier<RetreatResult> supplier;
+    if (!submerge || possible.size() > 1) {
+      title = "Retreat?";
+      supplier = () -> showRetreatDialog(message, possible);
+    } else {
+      title = "Submerge Subs?";
+      supplier = () -> showSubmergeDialog(message);
+    }
+    final Action action = getPlayerAction(title, supplier, future);
```
Comment (asvitkine) [inline game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: I think you missed the other part of my comment where I suggested not having this getPlayerAction() method and just inlining it here with the changes above, since it's only called once now.

File: game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java [hunk 3]
```diff
@@ -340,43 +349,66 @@ void notifyRetreat(final Collection<Unit> retreating) {
...
-            "Submerge Subs?",
-            e -> {
-              actionButton.setEnabled(false);
-              if (showSubmergeDialog(message, retreatTo)) {
-                actionButton.setAction(nullAction);
-                latch.countDown();
-              }
-              actionButton.setEnabled(true);
-            });
+    final CompletableFuture<Territory> future = new CompletableFuture<>();
```
Comment (asvitkine) [inline game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: Nit: Move future to be right above the "final Action action =" line.

File: game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java [hunk 4]
```diff
@@ -340,43 +349,66 @@ void notifyRetreat(final Collection<Unit> retreating) {
...
+      return future.get();
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } catch (final ExecutionException e) {
+      log.log(Level.INFO, "UiContext shut down before supplying result", e);
+    } finally {
+      mapPanel.getUiContext().removeActive(rejectionCallback);
+    }
+    return null;
+  }
```
Comment (asvitkine) [inline game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: Seems like this should live somewhere else and be generic to the return type. Since it's using UiContext, perhaps we can add the method on UiContext instead?

File: game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java (hunk 2 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: True, but if the method gets much longer code climate will complain.
That's why the method exists.

File: game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java (hunk 4 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java]: Valid point, but I saw this kind of pattern in a lot of places, so that would be a larger cleanup.
I plan renaming `removeActive` and `addActive` to `addShutdownHook` and `removeShutdownHook` and replace the `Active` interface with runnable in a follow-up PR.
I think the current status is definitely an improvement for now, but I agree that it can be improved further.
I'd move this method as part of this incoming PR then.
Any further refactoring is not on my priority list though, so it would be up to you in case you want to apply this pattern elsewhere.
